### PR TITLE
[LLHD] Process Lowering: support muxed signals

### PR DESCRIPTION
**Pull request by maerhart**
Canonicalization merges similar basic blocks and selects differing values via block arguments. This also includes values of LLHD signal type. The block-argument-to-mux pass then transforms them to muxes that need to be considered in process lowering.

### CODE REVIEW

File: lib/Dialect/LLHD/Transforms/ProcessLoweringPass.cpp [hunk 1]
```diff
@@ -60,10 +71,14 @@ void ProcessLoweringPass::runOnOperation() {
...
-            return wait.emitOpError(
-                "Process-lowering: The wait terminator is required to have "
-                "all probed signals as arguments!");
+          SmallVector<Value, 8> signals;
+          getAllMuxedSignals(prbOp.signal(), signals);
+          for (Value sig : signals) {
+            if (!llvm::is_contained(wait.obs(), sig)) {
+              return wait.emitOpError(
+                  "Process-lowering: The wait terminator is required to have "
+                  "all probed signals as arguments!");
```
Comment (fabianschuiki) [inline lib/Dialect/LLHD/Transforms/ProcessLoweringPass.cpp:80]: Just nitpicking: I think the primary error messages are supposed to start with a lowercase latter, something like:
```suggestion
                  "process-lowering: the wait terminator is required to have "
                  "all probed signals as arguments");
```

File: test/Dialect/LLHD/Transforms/processLoweringErrors.mlir [hunk 2]
```diff
@@ -61,3 +61,15 @@ llhd.proc @moreThanTwoBlocksNotAllowed() -> () {
...
+// -----
+
+llhd.proc @muxedSignal(%arg0 : !llhd.sig<i64>, %arg1 : !llhd.sig<i64>) -> () {
+  cf.br ^bb1
+^bb1:
+  %cond = hw.constant true
+  %sig = comb.mux %cond, %arg0, %arg1 : !llhd.sig<i64>
+  %0 = llhd.prb %sig : !llhd.sig<i64>
+  // expected-error @+1 {{Process-lowering: The wait terminator is required to have all probed signals as arguments!}}
+  llhd.wait (%arg0 : !llhd.sig<i64>), ^bb1
```
Comment (fabianschuiki) [inline test/Dialect/LLHD/Transforms/processLoweringErrors.mlir:74]: Cool, that is an excellent catch! Could this be a requirement that goes beyond multiplexers, where we'd have to follow all probed signals backwards through arbitrary operations on them to ensure they are part of the wait?

**Review by fabianschuiki — APPROVED** [review]
LGTM!

File: test/Dialect/LLHD/Transforms/processLoweringErrors.mlir (hunk 2 above)
Comment (maerhart) [reply test/Dialect/LLHD/Transforms/processLoweringErrors.mlir:74]: That's actually a good point! This depends on the LLHD code the frontend produces and the canonicalization patterns. I haven't seen another case using the current setup with moore, but only considering muxes is somewhat conservative and basically over-fitting to the kind of code moore produces. For example, another frontend could generate
```mlir
%hiSig = llhd.sig.extract %arg0 from %c16  // %arg0 is of type !llhd.sig<i32>
%prb = llhd.prb %hiSig
llhd.wait %hiSig   // both %hiSig and %arg0 are valid here
```
which indicates that we should follow backwards all op argument values of signal type and stop recursing on an argument as soon as it occurs in the observer list in llhd.wait.
We still don't have completeness with this approach though:
```mlir
%struct = hw.struct_create (a: %sig1, b: %sig2, c: %sig3)
%s1 = hw.struct_extract %struct[a]
%prb = llhd.prb %s1
llhd.wait %sig1  // both %sig1 and %s1 are valid here, but for %sig1 an error would be thrown
```
or
```mlir
%mux = comb.mux %true, %sig1, %sig2
%prb = llhd.prb %mux
llhd.wait %sig1
```
In the latter case it is valid to only observe `%sig1` because we statically know that `%sig2` will never be used. So I rely on folding and canonicalization to remove such cases.

Basically we can recurse on all operations where the union of all bits of all signal arguments is a superset of the bits of the result signal, but also have to check all intermediate SSA values of signal type.
In theory, we could have operations for which this does not hold:
```mlir
// this is invalid to be lowered, but would pass
%zext = llhd.sig.zext %sig // from 16 to 32 bits
llhd.connect %zext, %inoutarg // bidirectional connect allows other processes or entities to change the value driven onto the zero extended part
%prb = llhd.prb %zext
llhd.wait %sig
```
As we don't have any such operations in LLHD it should be safe to recurse for all operations, but still remains a best-effort pass that fails for some valid inputs (as usual when providing static guarantees).

Comment (fabianschuiki) [reply test/Dialect/LLHD/Transforms/processLoweringErrors.mlir:74]: Yeah good point about the best-effort angle of this. I actually like your approach of saying "this is what we currently get from our frontend" and then ensuring that this input works well. Because as you say, you'll always find inputs where this doesn't quite work, so trying to get _all the things_ to pass upfront seems like a futile approach. Maybe we can keep things as you have them right now since this solves an actually-occuring problem, and then revisit generalizing this to other ops once the need arises?

Comment (maerhart) [reply test/Dialect/LLHD/Transforms/processLoweringErrors.mlir]: I decided to generalize it a bit more, as you initially suggested, because it makes the pass more robust w.r.t. certain canonicalization patterns that could be added as well as changes in the lowering from Moore that could occur during the porting effort to CIRCT. This also allows us to remove the dependency on the comb dialect. In case this generalization is still not enough we should revisit it later as it would become too complicated to generalize further without any specific failing example.

File: lib/Dialect/LLHD/Transforms/ProcessLoweringPass.cpp (hunk 1 above)
Comment (maerhart) [reply lib/Dialect/LLHD/Transforms/ProcessLoweringPass.cpp]: I will address this in a follow-up commit to keep this functional addition separat from cleanup.

File: test/Dialect/LLHD/Transforms/processLoweringErrors.mlir (hunk 2 above)
Comment (fabianschuiki) [reply test/Dialect/LLHD/Transforms/processLoweringErrors.mlir]: That sounds like a great approach to me :+1:
